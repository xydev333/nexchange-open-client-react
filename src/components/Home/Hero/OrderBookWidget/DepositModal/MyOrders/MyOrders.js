import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { I18n } from 'react-i18next';
import axios from 'axios';
import config from 'Config';
import moment from 'moment';

import styles from './MyOrders.scss';


class MyOrders extends PureComponent {
  constructor(props) {
    super();

    this.state = {
        orders: [],
        expandedOrders: [],
        loading: true
    };

    this.fetchMyOrders = this.fetchMyOrders.bind(this);
    this.toggleOrder = this.toggleOrder.bind(this);
  }

  componentWillMount() {
    if(this.props.order){
      const expandedOrders = this.state.expandedOrders;
      expandedOrders.push(this.props.order.unique_reference);
      this.setState({expandedOrders});
    }

    this.fetchMyOrders();
    this.interval = setInterval(() => {
        this.fetchMyOrders();
    }, config.ORDER_BOOK_FETCH_INTERVAL)
  }

  componentWillUnmount() {
      clearInterval(this.interval);
  }

  toggleOrder(orderId) {
    let expandedOrders = this.state.expandedOrders;
    if(expandedOrders.indexOf(orderId) != -1){
      expandedOrders = _.remove(expandedOrders, function(n) {
        return n === orderId;
      });
    } else {
      expandedOrders.push(orderId);
    }
    this.setState({expandedOrders});
  }

  fetchMyOrders = () => {
    if(localStorage.limitOrderHistory){
        const limitOrderHistory = localStorage.limitOrderHistory.split(",");
        let orders = [];
        limitOrderHistory.forEach((orderId) => {
            const url = `${config.API_BASE_URL}/limit_order/${orderId.replace(/"/g,"")}/`;
            const request = axios.get(url);
            request
            .then(res => {
                orders.push(res.data);
                orders = _.sortBy(orders, function(order) {
                  return new Date(order.created_on);
                }).reverse();
                if(orders.length === limitOrderHistory.length){
                    this.setState({orders, loading: false});
                }
            })
            .catch(error => {
                console.log(error);
            });
        });
      }
  }

  componentWillUnmount() {
  }

  render() {
    return (
        <I18n ns="translations">
          {(t, i18n) => (
            <div className={`col-xs-12 ${styles.container}`}>
                <h4 className={styles.title}>{`My Orders`}</h4>
                { 
                <div className={`${styles.list}`}> 
                 {this.state.orders.map((order) => {
                  if(this.state.expandedOrders.indexOf(order.unique_reference) != -1){
                    console.log("HEREEEEE");
                    return (<div className={`${styles['list-item']}`} key={order.unique_reference}>
                            <p>{`Unique Reference: ${order.unique_reference}`}</p>
                            {order.book_status_name ?
                            <p>{`Status: ${order.book_status_name[0][1]}`}</p>
                            : null}
                            <p>{`Created: ${new moment(order.created_on).locale(`${i18n.language}`).fromNow()} `}</p>
                            <p>{`Withdraw Address: ${order.withdraw_address.address} 
                            (${order.withdraw_address.currency_code})`}</p>
                            <p>{`Deposit Address: ${order.deposit_address.address} 
                            (${order.deposit_address.currency_code})`}</p>
                            {order.status_name[0][1] === 'INITIAL' ? 
                            <span>{`In order to complete your order, send ${order.amount_base} 
                            ${order.deposit_address.currency_code}
                            to the deposit address`}</span> : null}
                            </div>);
                  } else {
                    return (<div className={`${styles['list-item']}`} key={order.unique_reference}>
                              <div className={`${styles['heading']}`}>
                                <div className={`col-xs-9 col-sm-9 col-md-6 col-lg-4`}>
                                  <h5>{`(${order.book_status_name[0][1]}) ${order.order_type == 1 ? 'Buy' : 'Sell'} 
                                  ${order.pair.base.code} - ${order.pair.quote.code}`}</h5>
                                  <span>{new moment(order.created_on).locale(`${i18n.language}`).fromNow()}</span>
                                </div>
                                <div className={`col-xs-3 col-sm-3 col-md-6 col-lg-8`}>
                                  <a onClick={() => this.toggleOrder(order.unique_reference)}>View details</a>
                                </div>
                              </div>
                            </div>)
                  }
                 })}
                </div>
                }
           </div>
          )}
        </I18n>
    );     
  }
}

const mapStateToProps = ({ }) => ({ });
const mapDispatchToProps = dispatch => bindActionCreators({ }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyOrders);
