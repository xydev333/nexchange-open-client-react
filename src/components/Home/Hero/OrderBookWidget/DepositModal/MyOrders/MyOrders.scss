.container {

    .title {
        font-size: 20px;
        font-weight: bold;
    }

    .list {
        max-height: 420px;
        overflow: auto;

        .list-item {
            padding: 5px 0px;

            .heading{
                display: flex;
                align-items: center;

                div {
                    display: flex;
                    flex-direction: column;

                    h5 {
                        margin: 0;
                        font-size: 16px;
                    }

                    span {
                        color: $grey;
                        font-size: 12px;
                    }
                }

                a {
                    cursor: pointer;
                }
            }
        }
    }
}