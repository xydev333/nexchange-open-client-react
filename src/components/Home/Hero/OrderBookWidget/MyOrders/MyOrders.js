import React, { PureComponent } from 'react';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import axios from 'axios';
import config from 'Config';
import { changeOrderBookValue } from 'Actions/index.js';

import MyOrdersCollapsed from './MyOrdersCollapsed/MyOrdersCollapsed';
import styles from './MyOrders.scss';
import MyOrdersExpanded from './MyOrdersExpanded/MyOrdersExpanded';


class MyOrders extends PureComponent {
  constructor(props) {
    super();

    this.state = {
        loading: true
    };

    this.fetchMyOrders = this.fetchMyOrders.bind(this);
  }

  UNSAFE_componentWillMount() {
    this.fetchMyOrders();
    this.interval = setInterval(() => {
        this.fetchMyOrders();
    }, config.ORDER_BOOK_FETCH_INTERVAL)
  }

  componentWillUnmount() {
      clearInterval(this.interval);
  }

  fetchMyOrders = () => {
    if(localStorage.orderHistory){
        try {
          let orderHistory = localStorage['orderHistory'];
          orderHistory = orderHistory ? JSON.parse(orderHistory).reverse() : [];          
          const limitOrderHistory = _.filter(orderHistory, { 'mode': 'LIMIT' });
          let orders = [];
          limitOrderHistory.forEach((order) => {
              const url = `${config.API_BASE_URL}/limit_order/${order.id.replace(/"/g,"")}/`;
              const request = axios.get(url);
              request
              .then(res => {
                  orders.push(res.data);
                  orders = _.sortBy(orders, function(order) {
                    return new Date(order.created_on);
                  }).reverse();
                  if(orders.length === limitOrderHistory.length){
                      const orderBook = this.props.orderBook;
                      orderBook.myOrders = orders;
                      this.props.changeOrderBookValue(orderBook)
                      this.setState({ loading: false });
                  }
              })
              .catch(error => {
                  console.log(error);
              });
          });
        } catch (e) {
          this.orderHistory = [];
        }
      }
  }

  render() {
    if(this.state.redirect) {
      return <Redirect push to="/?advanced=true&myorders=true" />;
    }
    
    if(this.props.expanded) {
        return (
          <div className={`col-xs-12 ${styles.wrapper}`}> 
            <MyOrdersExpanded 
              myOrders={this.props.orderBook.myOrders} 
              collapseMyOrders={this.props.collapseMyOrders}/>
          </div>
        );
      } else {
        return (
          <div className={`col-xs-12 col-sm-12 col-md-6 col-lg-4 ${styles.wrapper}`}> 
            <MyOrdersCollapsed 
              myOrders={this.props.orderBook.myOrders} 
              expandMyOrders={this.props.expandMyOrders} 
              goToMyOrders={() => this.setState({redirect: true})}
              shouldRedirect = {this.props.shouldRedirect}
              />
          </div>
        );
      }   
  }
}

const mapStateToProps = ({ orderBook }) => ({ orderBook });
const mapDispatchToProps = dispatch => bindActionCreators({ changeOrderBookValue }, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MyOrders);
