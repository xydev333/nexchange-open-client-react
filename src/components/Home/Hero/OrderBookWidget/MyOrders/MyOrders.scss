.container {

    .title {
        font-size: 16px;
        font-weight: bold;
    }

    .list {
        display: flex;
        flex-direction: column;
        max-height: 230px;
        overflow-x: hidden;
        overflow-y: auto;
        overflow-wrap: break-word;

        .list-item {
            padding: 5px 0px;

            .heading{

                div {
                    padding: 0;
                    display: flex;
                    flex-direction: column;

                    h5 {
                        margin: 0;
                        font-size: 14px;
                    }

                    span {
                        color: $grey;
                        font-size: 12px;
                    }
                }

                a {
                    font-size: 12px;
                    cursor: pointer;
                }
            }

            .details {
                padding: 0;
                padding-bottom: 10px;
                margin-bottom: 10px;
                border-bottom: 1px solid lighten( $gray, 10%);

                .field {
                    display: flex;
                    flex-direction: column;
                    margin: 5px 0px;
                    font-size: 12px;

                    .label {
                        font-weight: bold;    
                    }

                }

                .obs {
                    color: $gray;
                    font-size: 12px;
                }
            }
        }
    }
}