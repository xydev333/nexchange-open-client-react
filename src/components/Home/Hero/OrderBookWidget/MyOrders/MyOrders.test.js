import React from 'react';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MyOrders from './MyOrders';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('MyOrders', () => {
  const initialState = {
      orderBook: {
          myOrders: []
      }
  };
  let store, wrapShallow;

  beforeEach(() => {
    store = mockStore(initialState);
    localStorage.setItem('orderHistory', orderHistory);
    wrapShallow = shallow(<MyOrders store={store} />).dive();
  });

  it('renders correctly', () => {
    expect(wrapShallow).toMatchSnapshot();
  });
});

