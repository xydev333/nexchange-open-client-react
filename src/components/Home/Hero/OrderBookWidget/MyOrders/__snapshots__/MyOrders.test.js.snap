// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyOrders renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MyOrders
    changeOrderBookValue={[Function]}
    orderBook={
      Object {
        "myOrders": Array [],
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <MyOrdersCollapsed
        goToMyOrders={[Function]}
        myOrders={Array []}
      />,
      "className": "col-xs-12 col-sm-12 col-md-6 col-lg-4 wrapper",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "expandMyOrders": undefined,
        "goToMyOrders": [Function],
        "myOrders": Array [],
        "shouldRedirect": undefined,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <MyOrdersCollapsed
          goToMyOrders={[Function]}
          myOrders={Array []}
        />,
        "className": "col-xs-12 col-sm-12 col-md-6 col-lg-4 wrapper",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "expandMyOrders": undefined,
          "goToMyOrders": [Function],
          "myOrders": Array [],
          "shouldRedirect": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {
      "storeSubscription": undefined,
    },
    Symbol(enzyme.__providerValues__): Map {},
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
